zplugin-ros1/src/ros_to_zenoh_bridge/ros1_to_zenoh_bridge_impl.rs

```diff
@@ -0,0 +1,233 @@
+                    }
+                    Err(_) => todo!(),
+                }
+            }
+        }
+    }
+}
+
+struct RosToZenohBridge<RosStatusCallback, BridgeStatisticsCallback>
+where
+    RosStatusCallback: Fn(RosStatus),
+    BridgeStatisticsCallback: Fn(BridgeStatus),
+{
+    ros_status: RosStatus,
+    ros_status_callback: RosStatusCallback,
+
+    statistics_callback: BridgeStatisticsCallback,
+}
+
+impl<RosStatusCallback, BridgeStatisticsCallback>
+    RosToZenohBridge<RosStatusCallback, BridgeStatisticsCallback>
+where
+    RosStatusCallback: Fn(RosStatus),
+    BridgeStatisticsCallback: Fn(BridgeStatus),
+{
+    // PUBLIC
+    pub fn new(
+        ros_status_callback: RosStatusCallback,
+        statistics_callback: BridgeStatisticsCallback,
+    ) -> Self {
+        RosToZenohBridge {
+            ros_status: RosStatus::Unknown,
+            ros_status_callback,
+
+            statistics_callback,
+        }
+    }
+
+    pub async fn run(
+        &mut self,
+        ros1_client: Arc<Ros1Client>,
+        bridges: Arc<Mutex<BridgesStorage>>,
+        flag: Arc<AtomicBool>,
+    ) {
+        while flag.load(Relaxed) {
+            let cl = ros1_client.clone();
+            let ros1_state = async_std::task::spawn_blocking(move || {
+                topic_mapping::Ros1TopicMapping::topic_mapping(cl.as_ref())
+            })
+            .await;
+
+            debug!("ros state: {:#?}", ros1_state);
+
+            let mut locked = bridges.lock().await;
+            match ros1_state {
+                Ok(mut ros1_state_val) => {
+                    self.transit_ros_status(RosStatus::Ok);
+
+                    if locked.receive_ros1_state(&mut ros1_state_val).await {
+                        self.report_bridge_statistics(&locked).await;
+                        async_std::task::sleep(core::time::Duration::from_millis(100)).await;
+                    } else {
+                        self.report_bridge_statistics(&locked).await;
+                        async_std::task::sleep(core::time::Duration::from_millis(200)).await;
+                    }
+                }
+                Err(e) => {
+                    error!("Error reading ROS state: {}", e);
+
+                    self.transit_ros_status(RosStatus::Error);
+                    Self::cleanup(&mut locked);
+                    self.report_bridge_statistics(&locked).await;
+
+                    async_std::task::sleep(core::time::Duration::from_millis(500)).await;
+                }
+            }
+        }
+    }
+
+    // PRIVATE
+    fn transit_ros_status(&mut self, new_ros_status: RosStatus) {
+        if self.ros_status != new_ros_status {
+            self.ros_status = new_ros_status;
+            (self.ros_status_callback)(self.ros_status);
+        }
+    }
+
+    async fn report_bridge_statistics(&self, locked: &MutexGuard<'_, BridgesStorage>) {
```

Comment:
In this case there is no need to explicitly state that the input parameter is `&MutexGuard<'_, BridgesStorage>`.
You can express it as `&BridgesStorage`, Rust will automatically dereference the `MutexGuard` to the right type.